### PR TITLE
Uniforms in asm and some more

### PR DESCRIPTION
**Pull request by digetx**
One step closer to a cube-textured in assembly 🚀

### CODE REVIEW

File: src/libgrate/grate-asm.c [hunk 1]
```diff
@@ -398,6 +398,46 @@ struct grate_shader *grate_shader_parse_fragment_asm(const char *asm_txt)
...
+
+		switch (asm_fs_uniforms[i].used) {
+		case FS_UNIFORM_FX10_LOW:
+			cgc->symbols[cgc->num_symbols].type = GRATE_TYPE_FX10_LOW;
+			break;
+		case FS_UNIFORM_FX10_HIGH:
+			cgc->symbols[cgc->num_symbols].type = GRATE_TYPE_FX10_HIGH;
+			break;
+		case FS_UNIFORM_FP20:
+			cgc->symbols[cgc->num_symbols].type = GRATE_TYPE_FP20;
```
Comment (kusma) [inline src/libgrate/grate-asm.c]: This seems like you're setting up a symobol per scalar value. Is that so? If so, I don't think it's a great idea.

I think it'd be better if we used the types already defined in glsl_type, and augmented it with a precision-value. We could use location to device between GRATE_TYPE_FX10_LOW and GRATE_TYPE_FX10_HIGH, by requiring GRATE_TYPE_FP20 to be aligned to multiples of two, and letting the last bit of location as a differentiator for FX10...

**Comment by kusma** [comment]
Apart from the detail above, I think this is pretty nice :)

File: src/libgrate/grate-asm.c (hunk 1 above)
Comment (kusma) [reply src/libgrate/grate-asm.c]: By the way, it'd be totally fine to just support GLSL_TYPE_FLOAT as a first approach.

Comment (kusma) [reply src/libgrate/grate-asm.c]: And we might not want to keep the GLSL-prefixes here. There's no point in differentiating between GLSL-types and other types. IMO...

Comment (kusma) [reply src/libgrate/grate-asm.c]: The reason why I care about this, is that I don't want combinatory explosion here. Types like this explode very quickly...

In fact, we might want to have something like vector/matrix-sizes... Hmm, dunno, maybe I'm over-thinking this now...

Comment (digetx) [reply src/libgrate/grate-asm.c]: Thank you very much for the suggestions. BLOB seems encode fragment uniform type-position in the symbol location, maybe we can re-use it. However, the type enum seem differ from the vertex uniforms.

**Comment by digetx** [comment]
I corrected "libhost1x: add host1x_bo_flush()" commit a bit and and added couple minor updates for  	tools/assembler. Will do something about fragment uniforms ASAP 🤒

**Comment by digetx** [comment]
I squashed "grate: refactor grate_3d_ctx_set_fragment_uniform*()" and "grate: fragment asm: support uniforms" to "grate: support fragment uniforms". We are now use BLOB's symbol location encoding for fragment uniforms, so setting fragment uniform also works for cgc shaders. It is very similar to vertex uniforms, the difference is that BLOB supports only one type of fragment uniform - a vector and it's components enable mask. So a float uniform is a vector with 1 component enabled and mat4 is represented as four vec4's. Please take a look :)

**Review by kusma — APPROVED** [review]
Nice :)

**Pull request merged**